clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -330,8 +330,9 @@ void run(long now) {
             }
         }
 
-        long pollTimeout = sendProducerData(now);
-        client.poll(pollTimeout, now);
+        long currentTimeMs = time.milliseconds();
+        long pollTimeout = sendProducerData(currentTimeMs);
+        client.poll(pollTimeout, currentTimeMs);
     }
 
     private long sendProducerData(long now) {
```

Comment:
(This is a rewrite of the comment I thought I had left with my previous review)

With the changes in this patch, the path for the regular produce diverges from the transactional path: the former takes in a value of `now` from `runOnce` and uses it all the way through while the latter recomputes it. This is fine because the regular producer path has no blocking calls. It would be good add a comment explaining this at the top level.